### PR TITLE
#162948852 search functionality 

### PR DESCRIPTION
**Pull request by MaggieChege**
#### What does this PR do?
Implement functionality for the search for an article with a title, tag and by author

#### Description of Task to be completed?
A user can search for an  article with a certain title
A user can search for an  article by a certain author
A user can search for an  article with a certain tag

#### How should this be manually tested?

- Clone the repo locally

- cd ah-django-unchained
- git checkout ft-search-functionality-162948852
- get started with the app as from the README section. Follow the sample .env file supplied to obtain the right variables

use the following query params: tags, title, author to filter

**Screenshots**

**Search for article by author**

image

**Search for article by tag**

image

**Search for article by title**

image

**Search for an article by author, title and tag**

image

#### What are the relevant pivotal tracker stories?
[#162948852](https://www.pivotaltracker.com/story/show/162948852)

### CODE REVIEW

**Review by kwameasiago — APPROVED** [review]
Looks good

File: authors/apps/articles/tests/test_filter_articles.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+            "tagList": ["python"]
+        }
+
+    def register(self):
+        register = self.client.post(self.sign_up_url, self.sign_up_data,
+                                    format='json')
+        token = register.data['token']
+        return token
+
+    def test_filter_article_title(self):
```
Comment (kwameasiago) [inline authors/apps/articles/tests/test_filter_articles.py:84]: I think we should add a comment to explain what the test does

File: authors/apps/articles/filter.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
+from rest_framework.generics import ListAPIView
+from django_filters.rest_framework import DjangoFilterBackend
```
Comment (ignatiusukwuoma) [inline authors/apps/articles/filter.py]: DjangoFilterBackend should be available as filters.DjangoFilterBackend from the import on line 3

File: authors/apps/articles/urls.py [hunk 3]
```diff
@@ -8,6 +8,7 @@
                     CommentDelete, CommentsRetrieveUpdateDestroy,
                     ReadTime
                     )
+from . filter import FilterArticles
```
Comment (ignatiusukwuoma) [inline authors/apps/articles/urls.py]: There should't be space between . and filter

File: authors/apps/articles/tests/test_filter_articles.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        token = self.register()
+        # Create article
+        self.client.post(self.article_listcreate_url, self.article_data,
+                                    format='json',
+                                    HTTP_AUTHORIZATION='token {}'.format(token)
+                                    )
+        response = self.client.get(self.filter_title_url,
+                        format='json', HTTP_AUTHORIZATION='token {}'.format(token))
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        self.assertIn(self.article_data['title'],response.data['results'][0]['title'])
```
Comment (ignatiusukwuoma) [inline authors/apps/articles/tests/test_filter_articles.py]: But we only created one article, that article will definitely contain our title

File: authors/apps/articles/tests/test_filter_articles.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+    def test_filter_article_tag(self):
+        """Test that you can filter an article with a specific tag"""
+
+        token = self.register()
+        # Create article
+        self.client.post(self.article_listcreate_url, self.article_data,
```
Comment (ignatiusukwuoma) [inline authors/apps/articles/tests/test_filter_articles.py:135]: How about making a function that creates about 5 articles so you can properly test this.

File: authors/apps/articles/tests/test_filter_articles.py (hunk 4 above)
Comment (MaggieChege) [reply authors/apps/articles/tests/test_filter_articles.py]: I'll add another article.

File: authors/apps/articles/tests/test_filter_articles.py (hunk 5 above)
Comment (MaggieChege) [reply authors/apps/articles/tests/test_filter_articles.py:135]: Okay

File: authors/apps/articles/urls.py [hunk 6]
```diff
@@ -44,4 +45,7 @@
     path('articles/read/<slug>/',
          ReadTime.as_view(),
          name="articles-read"),
+
+    path('article/search/',FilterArticles.as_view(), name="search")
```
Comment (ignatiusukwuoma) [inline authors/apps/articles/urls.py]: should be `articles/`... Also, space after endpoint

**Comment by ignatiusukwuoma** [comment]
Run files through pycodestyle including tests

File: authors/apps/articles/tests/test_filter_articles.py [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+                                    )
+        self.client.post(self.article_listcreate_url, self.article_data1,
+                                    format='json',
+                                    HTTP_AUTHORIZATION='token {}'.format(token)
+                                    )
+        self.client.post(self.article_listcreate_url, self.article_data2,
+                                    format='json',
+                                    HTTP_AUTHORIZATION='token {}'.format(token)
+                                    )
+        response = self.client.get(self.filter_title_url,
```
Comment (ignatiusukwuoma) [inline authors/apps/articles/tests/test_filter_articles.py:120]: Do a normal `GET articles` request and assert that the title of the first result is not `self.article_data['title']`  before your filtered request. Same for other filters

File: authors/apps/articles/tests/test_filter_articles.py [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+                                    )
+        self.client.post(self.article_listcreate_url, self.article_data3,
+                                    format='json',
+                                    HTTP_AUTHORIZATION='token {}'.format(token)
+                                    )
+        response = self.client.get(self.filter_tag_url,
+                        format='json', HTTP_AUTHORIZATION='token {}'.format(token))
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        self.assertListEqual(self.article_data['tagList'],response.data['results'][0]['tagList'])
+
```
Comment (ignatiusukwuoma) [inline authors/apps/articles/tests/test_filter_articles.py:200]: Add test to filter by author

File: authors/apps/articles/filter.py [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+        return queryset.filter(tagList__name__icontains=value)
+
+    class Meta:
+        model = Article
+        fields = ['author', 'title', 'tags']
+
+
+class FilterArticles(ListAPIView):
+    serializer_class = ArticleSerializer
+    queryset = Article.objects.all()
```
Comment (ignatiusukwuoma) [inline authors/apps/articles/filter.py]: So we want to include deleted articles?

File: authors/apps/articles/tests/test_filter_articles.py [hunk 10]
```diff
@@ -0,0 +1,122 @@
+from django.urls import reverse
+from rest_framework.views import status
+from rest_framework.test import APITestCase, APIClient
+
+
+class FilterArticles(APITestCase):
+    def setUp(self):
+        self.client = APIClient()
+        self.sign_up_url = reverse('authentication:auth-register')
+        self.filter_title_url = 'http://127.0.0.1:8000/api/article/search/?title=Coding is cool'
```
Comment (ignatiusukwuoma) [inline authors/apps/articles/tests/test_filter_articles.py]: Spaces should't be in urls

File: authors/apps/articles/filter.py (hunk 9 above)
Comment (MaggieChege) [reply authors/apps/articles/filter.py]: Okay, fixed that

File: authors/apps/articles/tests/test_filter_articles.py (hunk 7 above)
Comment (MaggieChege) [reply authors/apps/articles/tests/test_filter_articles.py:120]: Done

File: authors/apps/articles/tests/test_filter_articles.py [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+                         )
+        self.client.post(self.article_listcreate_url, self.article_data3,
+                         format='json',
+                         HTTP_AUTHORIZATION='token {}'.format(token)
+                         )
+        all_articles = self.client.get(self.article_listcreate, format='json')
+        response = self.client.get(self.filter_tag_url,
+                                   format='json',
+                                   HTTP_AUTHORIZATION='token {}'.format(token))
+        self.assertNotEqual(self.article_data['title'], all_articles.data['results'][0]['title'])
```
Comment (ignatiusukwuoma) [inline authors/apps/articles/tests/test_filter_articles.py]: You are supposed to be testing tag here, why are you asserting for title

File: authors/apps/articles/tests/test_filter_articles.py [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+                         )
+        self.client.post(self.article_listcreate_url, self.article_data3,
+                         format='json',
+                         HTTP_AUTHORIZATION='token {}'.format(token)
+                         )
+        all_articles = self.client.get(self.article_listcreate, format='json')
+        response = self.client.get(self.filter_author_url,
+                                   format='json',
+                                   HTTP_AUTHORIZATION='token {}'.format(token))
+        self.assertNotEqual(self.article_data['title'], all_articles.data['results'][0])
```
Comment (ignatiusukwuoma) [inline authors/apps/articles/tests/test_filter_articles.py]: Why asserting for title?

File: authors/apps/articles/tests/test_filter_articles.py [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+
```
Comment (ignatiusukwuoma) [inline authors/apps/articles/tests/test_filter_articles.py:200]: Where are asserting for the author?

File: authors/apps/articles/tests/test_filter_articles.py [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+                         )
+        # get all articles
+        all_articles = self.client.get(self.article_listcreate, format='json')
+        response = self.client.get(self.filter_all,
+                                   format='json',
+                                   HTTP_AUTHORIZATION='token {}'.format(token))
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        self.assertNotEqual(self.article_data['title'], all_articles.data['results'][0])
+
+        self.assertIn(self.article_data['title'],
```
Comment (ignatiusukwuoma) [inline authors/apps/articles/tests/test_filter_articles.py]: SHouldn't this be an `assertEqual`?

File: authors/apps/articles/tests/test_filter_articles.py [hunk 15]
```diff
@@ -0,0 +1,159 @@
...
+                         )
+        self.client.post(self.article_listcreate_url, self.article_data2,
+                         format='json',
+                         HTTP_AUTHORIZATION='token {}'.format(token)
+                         )
+        all_articles = self.client.get(self.article_listcreate, format='json')
+        response = self.client.get(self.filter_title_url,
+                                   format='json',
+                                   HTTP_AUTHORIZATION='token {}'.format(token))
+        self.assertNotEqual(self.article_data['title'], all_articles.data['results'][0],)
```
Comment (ignatiusukwuoma) [inline authors/apps/articles/tests/test_filter_articles.py]: what are you comparing it with?

Comment (MaggieChege) [reply authors/apps/articles/tests/test_filter_articles.py]: Hadn't seen this.

**Comment by ignatiusukwuoma** [comment]
Remember to implement the part where... when I search for an author username/name or article title, it also shows me articles with similar author username/name or article title

File: authors/apps/articles/serializers.py [hunk 16]
```diff
@@ -48,7 +48,7 @@ class Meta:
 
     # Inserts the ID of the author of an article into the foreign key row
     def get_author(self, obj):
-        return obj.author.id
+        return obj.author.username
```
Comment (ignatiusukwuoma) [inline authors/apps/articles/serializers.py]: Why this change?

File: authors/apps/articles/serializers.py [hunk 17]
```diff
@@ -84,7 +84,7 @@ class Meta:
     """Get logged in user ID"""
 
     def get_author(self, obj):
-        return obj.author.id
+        return obj.author.username
```
Comment (ignatiusukwuoma) [inline authors/apps/articles/serializers.py]: ditto

File: authors/apps/articles/tests/test_filter_articles.py [hunk 18]
```diff
@@ -0,0 +1,162 @@
...
+                         format='json',
+                         HTTP_AUTHORIZATION='token {}'.format(token)
+                         )
+        all_articles = self.client.get(self.article_listcreate, format='json')
+        response = self.client.get(self.filter_title_url,
+                                   format='json',
+                                   HTTP_AUTHORIZATION='token {}'.format(token))
+        self.assertNotEqual(self.article_data['title'], all_articles.data['results'][0]['title'])
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        self.assertIn(self.article_data['title'],
```
Comment (ignatiusukwuoma) [inline authors/apps/articles/tests/test_filter_articles.py:127]: Use an assert equal if you expect them to be equal

File: authors/apps/articles/tests/test_filter_articles.py [hunk 19]
```diff
@@ -0,0 +1,162 @@
...
+                         )
+        self.client.post(self.article_listcreate_url, self.article_data3,
+                         format='json',
+                         HTTP_AUTHORIZATION='token {}'.format(token)
+                         )
+        all_articles = self.client.get(self.article_listcreate, format='json')
+        response = self.client.get(self.filter_author_url,
+                                   format='json',
+                                   HTTP_AUTHORIZATION='token {}'.format(token))
+        self.assertEqual(self.author_filter, all_articles.data['results'][0]['author'])
```
Comment (ignatiusukwuoma) [inline authors/apps/articles/tests/test_filter_articles.py]: Should these be equal?... all_articles is not a filtered query. I expect assertions to be similar to the others

File: authors/apps/articles/tests/test_filter_articles.py [hunk 20]
```diff
@@ -0,0 +1,162 @@
...
+        self.assertNotEqual(self.author_filter2, all_articles.data['results'][0]['author'])
```
Comment (ignatiusukwuoma) [inline authors/apps/articles/tests/test_filter_articles.py]: Nice one

File: authors/apps/articles/filter.py [hunk 21]
```diff
@@ -0,0 +1,25 @@
+from django_filters import rest_framework as filters
+from rest_framework.generics import ListAPIView
+from .serializers import ArticleSerializer
+from .models import Article
+
+
+class ArticleFilter(filters.FilterSet):
+    author = filters.CharFilter(field_name='author__username',
```
Comment (ignatiusukwuoma) [inline authors/apps/articles/filter.py:8]: Would it be better to use the author's name?

File: authors/apps/articles/serializers.py (hunk 16 above)
Comment (MaggieChege) [reply authors/apps/articles/serializers.py]: So that we can return the authors username. This makes it easier for the tests.

File: authors/apps/articles/tests/test_filter_articles.py (hunk 19 above)
Comment (MaggieChege) [reply authors/apps/articles/tests/test_filter_articles.py]: Okay, gotten rid of it.

File: authors/apps/articles/filter.py (hunk 21 above)
Comment (MaggieChege) [reply authors/apps/articles/filter.py:8]: I think so, the other option would be the authors ID.
